sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala

```diff
@@ -121,7 +121,12 @@ private[hive] class SparkExecuteStatementOperation(
         result.toLocalIterator.asScala
       } else {
         if (resultList.isEmpty) {
-          resultList = Some(result.collect())
+          val limited = sqlContext.getConf(SQLConf.THRIFTSERVER_RESULT_LIMIT.key).toInt
+          resultList = if (limited > 0) {
+            Some(result.take(result.queryExecution.analyzed.maxRows.getOrElse[Long](limited).toInt))
```

Comment:
I am afraid you might get an unexpected outcome. `maxRows` was not introduced for this purpose. Sometimes, the `maxRows`'s default value is None. It might not always be popped up to the top query node.